### ISSUE TITLE
Oauth token refresh can trigger rate limiting and log errors

### ISSUE THREAD

**Issue by katz12** [camunda/camunda#13832]
**Describe the bug**

I have observed that my Zeebe client app will periodically emit error logs with `Failed while fetching credentials:  exception=java.io.IOException: Failed while requesting access token with status code 429 and message Too Many Requests.`

I don't see any other adverse effects besides the error logs, and the application seems to be able to continue to handle jobs afterwards.

**To Reproduce**

This is difficult to reproduce specifically, and the stack trace contains only Camunda code. It seems like this is coming from `JobPoller` and occurs when it is time to refresh the Zeebe client access token.

Perhaps something about the way that we register job handlers could be contributing to this. If we are hitting rate limits, then maybe multiple pollers are trying to refresh the access token at the same time.
```
    // WorkflowWorker and WorkerHandler are classes that our code uses to manage our various job handlers

    void register(WorkflowWorker<?>... workers) {
        Stream.of(workers).forEach(this::registerWorker);
    }

    private void registerWorker(WorkflowWorker<?> worker) {
        String jobType = worker.getJobType();
        JobHandler jobHandler = new WorkerHandler<>(worker, clientId);
        zeebeClient.newWorker()
                .jobType(jobType)
                .handler(jobHandler)
                .open();
    }
```

**Expected behavior**

I see two possible expected behaviors depending on what exactly is occurring.

1. If it is the case that these rate limit errors are causing job polling to fail, then I would expect that the token refresh implements an exponential backoff to navigate the rate limiting.
    - I saw a similar issue and fix in the community managed NodeJS client: https://github.com/camunda-community-hub/zeebe-client-node-js/issues/301
    - There is also a note on that issue that refers to the Java client implementing backoff in JobPoller, so it is possible that this error is not causing the poll to fail.

2. If it is the case that it is not causing job polling to fail, then I would expect that OauthCredentialsProvider no longer logs errors when encountering a 429.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
Failed while fetching credentials: 
java.io.IOException: Failed while requesting access token with status code 429 and message Too Many Requests.
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.fetchCredentials(OAuthCredentialsProvider.java:183)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.refreshCredentials(OAuthCredentialsProvider.java:132)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.shouldRetryRequest(OAuthCredentialsProvider.java:100)
	at io.camunda.zeebe.client.impl.worker.JobPoller.onError(JobPoller.java:116)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:487)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

</p>
</details>

**Environment:**
- OS: Linux
- Zeebe Version: 8.2.11
- Configuration: _unsure_

Related:
- https://jira.camunda.com/browse/SUPPORT-18518

**Comment by korthout** [comment]
Thanks for reporting @katz12 🙇 

@camunda/zeebe-process-automation (ZPA) triage:
- typically, the credentials provider's cache should avoid you from hitting such a 429
- https://github.com/camunda-community-hub/zeebe-client-node-js/issues/301 is about aligning the nodejs client with the exponential backoff that is already implemented by the java client's job poller. It does not appear to be related to this specific issue.

@katz12 could you please share more details about your setup/configuration so we can understand how this could happen?
- e.g. number of job workers configured in this client, `pollInterval`, and other client configurations, as well as how many job workers are you running in total (i.e. replicas)
- we're also curious whether you see any other logging, perhaps you could try lower the logging level to DEBUG to find anything out of the ordinary related to caching the credentials
- lastly, it would help us to understand whether you're on C8 SaaS or Self-Managed.

**Comment by katz12** [comment]
Hi @korthout thank you for the reply!

> typically, the credentials provider's cache should avoid you from hitting such a 429

This is great to hear! I did not mean to imply that the fix would be the same as the nodejs issue I linked but to suggest that this type of issue would likely appear during token refresh unless there was some backoff implemented, which it seems like there is.

> could you please share more details about your setup/configuration

**number of job workers**: 13
**pollInterval**: whatever is the default (which seems to be 100ms?)
**number of total workers**: 130 if I understand your question correctly. This is occurring mostly in our test environment where we have 10 different instances of the application that registers the 13 job workers running at the moment.

> whether you see any other logging

I do see a lot of WARN logs that occur at the same time with `UNAUTHENTICATED: HTTP status code 401` which makes a lot of sense since the token has expired. But those logs are occurring much more frequently than the errors, on the order of about 3 per second over about 3 minutes.

<details><summary>Full Stacktrace</summary>
 <p>

```
message=Failed to activate jobs for worker default and job type dpt:initializeProcess exception=io.grpc.StatusRuntimeException: UNAUTHENTICATED: HTTP status code 401
invalid content-type: text/html
headers: Metadata(:status=401,date=Wed, 16 Aug 2023 18:29:05 GMT,content-type=text/html,strict-transport-security=max-age=63072000; includeSubDomains,content-length=172)
DATA-----------------------------
<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></center>
<hr><center>nginx</center>
</body>
</html>

	at io.grpc.Status.asRuntimeException(Status.java:539)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:487)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:576)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:757)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:736)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
 level=WARN service=workflow-worker logger=io.camunda.zeebe.client.job.poller thread=grpc-default-executor-39 timestamp=2023-08-16T18:29:06.107Z
```

</p>
</details>

I can try enabling DEBUG logging next week and see if anything comes up.

> it would help us to understand whether you're on C8 SaaS or Self-Managed

C8 SaaS

**Comment by katz12** [comment]
If it is the case that the backoff is doing its job to navigate the rate limiting and eventually acquire the token, would it be possible to omit or downgrade the logging to WARN when encountering a 429?

**Comment by korthout** [comment]
Thanks for the additional input @katz12 

I haven't yet had time to look into this again. I've moved it into the team's inbox for triage and discussion.

**Comment by korthout** [comment]
@katz12 Since you're using C8 SaaS, the team recommends raising this to [our dedicated support department](https://docs.camunda.io/docs/components/console/console-troubleshooting/feedback-and-support/).

**Comment by megglos** [comment]
moving back to inbox and flagged as support related, let's validate whether we really consider this a bug or FR and decide on how to handle it + provide a timeline.

**Comment by abbasadel** [comment]
ZPA triage:
- As discussed on the support ticket, a support engineer will take over the ticket to better support the customer.
- Moving the issue to the backlog with 'Later' priority until get get the support engineer's feedback

**Comment by nicpuppa** [comment]
We faced another support case (internal: [22115](https://jira.camunda.com/browse/SUPPORT-22115) ) that seems related to the same problem.

The customer is facing a high load of request towards keycloak (when DB restarts), that make their client application not working anymore

The customer complains about the `OAuthCredentialsProvider#shouldRetryRequest` method implementation, that seems to not handle properly the retry mechanism.

I'm reopening this issue again so we can re-triage it with more information. 

This might be an already known issue in the credential cache implementation, as suggested [here](https://camunda.slack.com/archives/C037W9NMATG/p1718611057712049?thread_ts=1718610086.199119&cid=C037W9NMATG) by @koevskinikola 

It might be a valid candidate for the next mob-programming session

**Comment by mustafadagher** [comment]
ZPA Triage:
- This has been reopened due to a support case
- No reproduction steps were provided by customer
- Might be a valid candidate for the next mob-programming session
- No workaround, so severity is high

**Comment by koevskinikola** [comment]
ZPA mob programming analysis:
- The [OAuthCredentialsCache](https://github.com/camunda/camunda/blob/ed8a807969f040fdb55b575a9e31fef957fd8d5e/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L40) class is responsible for caching `ZeebeClientCredentials`. For this purpose, it uses a `HashMap` as [the in-memory cache](https://github.com/camunda/camunda/blob/ed8a807969f040fdb55b575a9e31fef957fd8d5e/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L48), and a `File` cache for more [persistent storing](https://github.com/camunda/camunda/blob/ed8a807969f040fdb55b575a9e31fef957fd8d5e/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L49) of the cached credentials. 
- The `OAuthCredentialsCache` class is marked as `ThreadSafe`. We can see that there are several `synchronized` methods in the `OAuthCredentialsCache` class which invoke methods of the (in-memory) `audience` HashMap. 
  - However, the operations on the HashMap are not thread-safe. Specifically, we have a suspicion that the following `OAuthCredentialsCache` methods may bring the `audience` object in an invalid state if called concurrently
    - [`OAuthCredentialsCache#readCache`](https://github.com/camunda/camunda/blob/ed8a807969f040fdb55b575a9e31fef957fd8d5e/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L56-L66)
    - [`OAuthCredentialsCache#put`](https://github.com/camunda/camunda/blob/ed8a807969f040fdb55b575a9e31fef957fd8d5e/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L117-L121)
  - We found that [there is a test case](https://github.com/camunda/camunda/blob/ed8a807969f040fdb55b575a9e31fef957fd8d5e/clients/java/src/test/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCacheTest.java#L231-L277) to ensure thread safety of the `OAuthCredentialsCache` but the test case doesn't seem to cover different cases of concurrent modification of the `audience` HashMap.
- The `OAuthCredentialsCache` is currently used by the `OAuthCredentialsProvider`, which in turn is used by the Zeebe Java client, and the Zeebe Spring SDK. There are currently no reproduction steps for this issue, but we don't see any occurrences of this issue on Spring SDK versions that use the community-built mechanism. As a result, we can switch to that mechanism in the Zeebe Spring SDK until we confirm the issue with the `OAuthCredentialsCache`,

**Comment by korthout** [comment]
I've had another look at the topic together with @remcowesterhoud.

We're were able to confirm an endless retry loop under special lab conditions. However, whether the user faced these specific conditions is unclear as we're unable to reproduce the problem in a running system. 

We've confirmed that a 429 would stop this endless loop, which aligns with what the user sees.

I'll open a PR to limit the number of tries per command to a total of 3 (i.e. 2 retries). 

I've also opened a PR (#20065) to resolve threadsafety issues in the OAuth credentials cache. These may have been a root cause for the endless retry, although we've not been able to confirm this with certainty. Resolving these issues will at least improve our confidence in this part of the code, and possibly resolve the user's endless retry issues as well.
